tests/multigrid/dof_06.cc

```diff
@@ -0,0 +1,172 @@
+  dof_handler.distribute_mg_dofs();
+
+  // Try to assemble MGConstrainedDofs
+  MGConstrainedDoFs constrained_dofs;
+  constrained_dofs.initialize(dof_handler);
+
+  // Report which DIM is ok
+  std::cout << "--- DIM " << dim << " OK ---" << std::endl;
+
+  // Debug printouts:
+  for (unsigned int level = 0; level < tria.n_global_levels(); ++level)
+    {
+      std::cout << "Level " << level << ": " << std::endl;
+
+      // Print faces that connect to coarser DoF
+      std::cout << "  Faces neighboring coarser cells:" << std::endl;
+      for (typename DoFHandler<dim>::cell_iterator level_cell =
+             dof_handler.begin(level);
+           level_cell != dof_handler.end(level);
+           ++level_cell)
+        {
+          // Iterate over all faces
+          for (unsigned int i_face = 0; i_face < 2 * dim; ++i_face)
+            {
+              if ((level_cell->at_boundary(i_face) &&
+                   !level_cell->has_periodic_neighbor(i_face)) ||
+                  level_cell->neighbor_or_periodic_neighbor(i_face)->level() ==
+                    level_cell->level())
+                continue;
+
+              // Get face
+              auto face = level_cell->face(i_face);
+
+              std::cout << "    ";
+              print_face(face, level);
+
+              // Print whether the neighbor is periodic
+              if (level_cell->has_periodic_neighbor(i_face))
+                std::cout << " via periodic boundary" << std::endl;
+              else
+                std::cout << " via internal face" << std::endl;
+            }
+        }
+
+      // Print hanging node DoFs
+      std::cout << "  Refinement edge DoFs: ";
+      constrained_dofs.get_refinement_edge_indices(level).print(std::cout);
+      std::cout << std::endl;
+
+      // Print faces that have periodic neighbors
+      std::cout << "  Faces having periodic neighbors of same level:"
+                << std::endl;
+      for (typename DoFHandler<dim>::cell_iterator level_cell =
+             dof_handler.begin(level);
+           level_cell != dof_handler.end(level);
+           ++level_cell)
+        {
+          // Iterate over all faces
+          for (unsigned int i_face = 0; i_face < 2 * dim; ++i_face)
+            {
+              if (!level_cell->has_periodic_neighbor(i_face) ||
+                  level_cell->periodic_neighbor(i_face)->level() !=
+                    level_cell->level())
+                continue;
+
+              typename DoFHandler<dim>::cell_iterator neighbor_cell =
+                level_cell->periodic_neighbor(i_face);
+
+              // Each side only once
+              if (level_cell->index() > neighbor_cell->index())
+                continue;
+
+              unsigned int neighbor_i_face =
+                level_cell->periodic_neighbor_face_no(i_face);
+
+              // Get face
+              auto face          = level_cell->face(i_face);
+              auto neighbor_face = neighbor_cell->face(neighbor_i_face);
```

Comment:
`const auto face` and `const auto neighbor_face`.

Reply:
Done.